### PR TITLE
IF: Further work of validating QC claim 

### LINKED ISSUE #2071: IF: Reject blocks with unsupported claims of last QC in block header extension
The block header extension for instant finality makes a claim of the block number of the last block in its ancestry for which it has seen a QC. It also makes a claim as to whether this QC was strong or weak.

The node should not simply trust these claims. It needs to validate them somehow.

One way of validating this is if the claim is the same as the claim in the prior block that was already validated. This can be seen by looking at the header extension within the header stored in the `block_header_state` that the block builds upon.

But often the claim will be new. And in this case, to verify the claim, the node must verify that a valid QC (of the appropriate type) exists on the claimed block. Note on appropriate type: if the claim is that a strong QC exists, the node must verify it has a strong QC on the block; if the claim is that a weak QC exists, the node must verify it has some QC (strong or weak) on the block.

The block should include the necessary QC in its block extension. So one way to carry out this verification process is to first integrate the QC provided from that block (after checking it is valid) into the `block_state` associated with the claimed block. This should be done anyway because the QC may be valuable information to keep around even if the received (new) block is abandoned anyway due to a fork reorganization or because we later realize the received block is invalid and needs to be thrown away.

Note: When integrating the QC into the `block_state` of the claimed block, it may provide the information/proof needed to allow LIB to advance. LIB should be advanced at this point (or shortly after) without having to wait for a new block to be built/completed or for a new vote message to be received.

Then validating the claim in the received block always involves looking up the `block_state` for the claimed block and ensuring it has a valid QC on it of the appropriate type.

A note on integrating the QC into the `block_state`:
The `block_state` may have a pending quorum certificate that the node is collecting votes for. This should not be touched when integrating a received QC into the `block_state`; there is a separate optional field for the `valid_qc`. The received QC should be placed in the field for the `valid_qc` but only if is better than any existing one in `valid_qc`. A strong valid QC is better than a weak valid QC. For now, do not bother comparing two different valid QC of the same type; they can be treated as equally good.
When observing the QC available for a particular block, we want to pick the best one available. This means that an accessor function `get_best_qc` should compare the pending quorum certificate (assuming it is valid) against the `valid_qc` and choose the better of the two (again better in this sense can just consider strong vs weak).

### PR DESCRIPTION
**Pull request by linh2931**
This PR resolved additional comments by @heifner on https://github.com/AntelopeIO/leap/pull/2120. 

Thanks Kevin!

* moved `integrate_received_qc_to_block` to `push_block` so that it is thread safe, 
* searched the claimed QC block on the branch of `id` not head id, 
* and added block number to all error messages for easier identifying offending block.

Resolved #2071

### CODE REVIEW

**Review by heifner — CHANGES_REQUESTED** [review]

File: libraries/chain/controller.cpp [hunk 1]
```diff
@@ -3029,6 +3043,7 @@ struct controller_impl {
       return block_handle{bsp};
    }
 
+   // Called in push_block, thread safe
```
[the remark this answers is not in the document]
Comment (linh2931) [reply libraries/chain/controller.cpp:3046]: `fork_db_fetch_bsp_by_num( id, qc_claim.last_qc_block_num )` seems not work. When running `nodeos_lib_test.py` I got an exception `Block state was not found in forkdb for last_qc_block_num 60. Block number: 61` after calling it.

[the remark this answers is not in the document]
Comment (heifner) [reply libraries/chain/controller.cpp:3046]: https://github.com/AntelopeIO/leap/pull/2139/files#r1466731733

File: libraries/chain/controller.cpp [hunk 2]
```diff
@@ -1013,6 +1014,19 @@ struct controller_impl {
       );
    }
 
+   // search on the branch of given id
+   block_state_ptr fork_db_fetch_bsp_by_num(const block_id_type& id, uint32_t block_num) const {
+      return fork_db.apply<block_state_ptr>(
+         overloaded{
+            [](const fork_database_legacy_t&) -> block_state_ptr { return nullptr; },
```
Comment (arhag) [inline libraries/chain/controller.cpp:1021]: I suppose we don't use it. But we could do a `search_on_branch` with the legacy forkdb as well.

Comment (heifner) [reply libraries/chain/controller.cpp:1021]: Not that returns the same type.

Comment (arhag) [reply libraries/chain/controller.cpp:1021]: Oh right. Never mind then.

File: libraries/chain/controller.cpp [hunk 3]
```diff
@@ -3180,21 +3202,21 @@ struct controller_impl {
...
                   block_validate_exception,
-                  "QC is_strong (${s1}) in block extension does not match is_last_qc_strong (${22}) in header extension",
-                  ("s1", qc_proof.qc.is_strong())("s2", qc_claim.is_last_qc_strong) );
+                  "QC is_strong (${s1}) in block extension does not match is_last_qc_strong (${22}) in header extension. Block number: ${b}",
+                  ("s1", qc_proof.qc.is_strong())("s2", qc_claim.is_last_qc_strong)("b", b->block_num()) );
 
-      // find the claimed block's block state
-      auto bsp = fork_db_fetch_bsp_by_num( qc_claim.last_qc_block_num );
+      // find the claimed block's block state on branch of id
+      auto bsp = fork_db_fetch_bsp_by_num( id, qc_claim.last_qc_block_num );
```
[the remark this answers is not in the document]
Comment (linh2931) [reply libraries/chain/controller.cpp:3215]: Thanks!

**Review by heifner — APPROVED** [review]

**Review by arhag — APPROVED** [review]

**Comment by ericpassmore** [comment]
Note:start
group: IF
category: INTERNALS
summary: Further enhancement to checking claim logic by review ancestor blocks. Along with code changes, includes better logging, and code comments.
Note: end

**Pull request merged**